features/topology-map/plugins/org.opennms.features.topology.plugins.topo.asset/src/main/java/org/opennms/features/topology/plugins/topo/asset/AssetGraphDefinitionRepositoryImpl.java

```diff
@@ -0,0 +1,244 @@
+		String propId=DEFINITION_PREFIX+providerId;
+		Configuration config;
+		try {
+			config = configurationAdmin.getConfiguration(persistentId);
+			Dictionary<String, Object> props = config.getProperties();
+
+			// if null, there is no configuration
+			if (props == null) return false;
+
+			String graphDefinitionUri = (String) props.get(propId);
+			if(graphDefinitionUri == null) return false;
+
+			return true;
+
+		} catch (Exception e) {
+			throw new RuntimeException("problem checking if definition exists  "+propId+ " from "+persistentId+".cfg",e);
+		}
+
+	}
+
+	/* (non-Javadoc)
+	 * @see org.opennms.features.topology.plugins.topo.asset.AssetGraphDefinitionRepository#getAllConfigDefinitions()
+	 */
+	@Override
+	public Map<String,GeneratorConfig> getAllConfigDefinitions(){
+		Map<String,GeneratorConfig> generatorConfigs = new LinkedHashMap<String,GeneratorConfig>();
+
+		try {
+			Configuration config = configurationAdmin.getConfiguration(persistentId);
+			Dictionary<String, Object> props = config.getProperties();
+
+			// if null, there are no configurations
+			if (props == null) return generatorConfigs;
+
+			Enumeration<String> keys = props.keys();
+			while(keys.hasMoreElements()){
+				String propId = keys.nextElement();
+				if(propId.startsWith(DEFINITION_PREFIX)){
+					String providerId=propId.replace(DEFINITION_PREFIX,"");
+					String graphDefinitionUri = (String) props.get(propId);
+					GeneratorConfig generatorConfig = new GeneratorConfigBuilder()
+					.withGraphDefinitionUri(graphDefinitionUri)
+					.build();
+					generatorConfigs.put(providerId, generatorConfig);
+				}
+			}
+
+		} catch (Exception e) {
+			throw new RuntimeException("problem loading all graph definitions from "+persistentId+".cfg",e);
+		}
+		return generatorConfigs;
+	}
+
+	/* (non-Javadoc)
+	 * @see org.opennms.features.topology.plugins.topo.asset.AssetGraphDefinitionRepository#removeConfigDefinition(java.lang.String)
+	 */
+	@Override
+	public void removeConfigDefinition(String providerId){
+		String propId=DEFINITION_PREFIX+providerId;
+
+		Configuration config;
+		try {
+			config = configurationAdmin.getConfiguration(persistentId);
+			Dictionary<String, Object> props = config.getProperties();
+
+			// if null, there is no configuration
+			if (props == null) return;
+
+			props.remove(propId);
+			config.update(props);
+			
+			// TODO WORK AROUND for KARAF 2.4 ISSUE writing directly to file instead of through ConfigAdmin service 
+			// Work around here because ConfigAdmin service can only add properties and not remove them from a cfg file
+			// This might be solvable by setting config type append=false in features.xml. 
+			// But this cannot be parsed by opennms's maven-features-plugin
+			// see line 296 org.apache.karaf.features.internal.service.FeatureConfigInstaller
+			// the FILEINSTALL_FILE_NAME is set by the <cm:property-placeholder in blueprint.xml
+
+			String cfgFile = (String) props.get(FILEINSTALL_FILE_NAME);
```

Comment:
The usage of the ConfigAdmin seems to be not the right tool. 
Therefore I would prefer to use JAXB to save/load asset configuration on disk as a xml file.
This would remove some code created to make it work with ConfigAdmin.